// @reclaimprotocol/attestor-core/lib/server/utils/apm.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getApm = getApm;
const elastic_apm_node_1 = __importDefault(require("elastic-apm-node"));
const env_1 = require("../../utils/env");
const logger_1 = require("../../utils/logger");
let apm;
/**
 * Initialises the APM agent if required,
 * and returns it.
 * If ELASTIC_APM_SERVER_URL & ELASTIC_APM_SECRET_TOKEN
 * are not set will return undefined
 *
 * Utilises the standard env variables mentioned
 * here: https://www.elastic.co/guide/en/apm/agent/nodejs/current/custom-stack.html#custom-stack-advanced-configuration
 */
function getApm() {
    if (!(0, env_1.getEnvVariable)('ELASTIC_APM_SERVER_URL')
        || !(0, env_1.getEnvVariable)('ELASTIC_APM_SECRET_TOKEN')) {
        logger_1.logger.info('ELASTIC_APM_SERVER_URL or ELASTIC_APM_SECRET_TOKEN not found'
            + ' in env, APM agent not initialised');
        return undefined;
    }
    if (!apm) {
        const sampleRate = +((0, env_1.getEnvVariable)('ELASTIC_APM_SAMPLE_RATE')
            || '0.1');
        apm = elastic_apm_node_1.default.start({
            serviceName: 'reclaim_attestor',
            serviceVersion: '4.0.0',
            transactionSampleRate: sampleRate,
            instrumentIncomingHTTPRequests: true,
            usePathAsTransactionName: true,
            instrument: true,
            captureHeaders: true,
        });
        logger_1.logger.info('initialised APM agent');
    }
    return apm;
}
