// @reclaimprotocol/attestor-core/lib/utils/bgp-listener.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBgpListener = createBgpListener;
const ip_cidr_1 = __importDefault(require("ip-cidr"));
const config_1 = require("../config");
const ws_1 = require("../utils/ws");
const ANNOUNCEMENT_OVERLAP = 'announcement-overlap';
class BGPAnnouncementOverlapEvent extends Event {
    constructor(data) {
        super(ANNOUNCEMENT_OVERLAP);
        this.data = data;
    }
}
/**
 * Listens for BGP announcements and emits events whenever
 * an announcement overlaps with a target IP.
 */
function createBgpListener(logger) {
    let ws;
    let closed = false;
    const targetIps = new Set();
    const eventTarget = new EventTarget();
    openWs();
    return {
        onOverlap(ips, callback) {
            for (const ip of ips) {
                targetIps.add(ip);
            }
            eventTarget.addEventListener(ANNOUNCEMENT_OVERLAP, _callback);
            return () => {
                for (const ip of ips) {
                    targetIps.delete(ip);
                }
                eventTarget.removeEventListener(ANNOUNCEMENT_OVERLAP, _callback);
            };
            function _callback(event) {
                callback(event.data);
            }
        },
        close() {
            ws.onclose = null;
            ws.onerror = null;
            ws.close();
            closed = true;
        }
    };
    function openWs() {
        logger.debug('connecting to BGP websocket');
        ws = (0, ws_1.makeWebSocket)(config_1.BGP_WS_URL);
        ws.onopen = onOpen;
        ws.onerror = (err) => onClose(err);
        ws.onclose = () => onClose(new Error('Unexpected close'));
        ws.onmessage = ({ data }) => {
            const str = typeof data === 'string' ? data : data.toString();
            try {
                onMessage(str);
            }
            catch (err) {
                logger.error({ data, err }, 'error processing BGP message');
            }
        };
    }
    function onOpen() {
        const subscriptionMessage = {
            type: 'ris_subscribe',
            data: {
                type: 'UPDATE',
            },
        };
        ws.send(JSON.stringify(subscriptionMessage));
        logger.info('connected to BGP websocket');
    }
    function onClose(err) {
        if (closed) {
            return;
        }
        logger.info({ err }, 'BGP websocket closed');
        if (!err) {
            return;
        }
        logger.info('reconnecting to BGP websocket');
        openWs();
    }
    function onMessage(message) {
        var _a, _b;
        const data = JSON.parse(message);
        const announcements = (_a = data === null || data === void 0 ? void 0 : data.data) === null || _a === void 0 ? void 0 : _a.announcements;
        logger.trace({ data }, 'got BGP update');
        if (!Array.isArray(announcements)) {
            return;
        }
        const asPath = (_b = data === null || data === void 0 ? void 0 : data.data) === null || _b === void 0 ? void 0 : _b.path;
        for (const announcement of announcements) {
            const prefixes = announcement === null || announcement === void 0 ? void 0 : announcement.prefixes;
            const nextHop = announcement === null || announcement === void 0 ? void 0 : announcement['next_hop'];
            const hasPrefixes = (prefixes === null || prefixes === void 0 ? void 0 : prefixes.length) && (nextHop || asPath);
            if (!hasPrefixes) {
                return;
            }
            for (const prefix of prefixes) {
                if (!overlapsTargetIps(prefix)) {
                    continue;
                }
                // emit event
                eventTarget.dispatchEvent(new BGPAnnouncementOverlapEvent({ prefix }));
            }
        }
    }
    function overlapsTargetIps(prefix) {
        // ignore all prefixes that end with /0
        if (prefix.endsWith('/0')) {
            return false;
        }
        const cidr = new ip_cidr_1.default(prefix);
        for (const ip of targetIps) {
            if (cidr.contains(ip)) {
                return true;
            }
        }
        return false;
    }
}
