// @reclaimprotocol/attestor-core/lib/utils/connection-state-machine.js
"use strict";
/**
 * Connection state machine for managing atomic state transitions.
 * Prevents race conditions in connection state management.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionStateMachine = exports.ConnectionState = void 0;
var ConnectionState;
(function (ConnectionState) {
    ConnectionState["CONNECTING"] = "connecting";
    ConnectionState["CONNECTED"] = "connected";
    ConnectionState["CLOSING"] = "closing";
    ConnectionState["CLOSED"] = "closed";
})(ConnectionState || (exports.ConnectionState = ConnectionState = {}));
class ConnectionStateMachine {
    constructor(initialState = ConnectionState.CONNECTING) {
        this.listeners = [];
        this.transitionLock = null;
        this.currentState = initialState;
    }
    /**
     * Get current state.
     */
    getState() {
        return this.currentState;
    }
    /**
     * Check if the connection is in any of the specified states.
     */
    isInState(...states) {
        return states.includes(this.currentState);
    }
    /**
     * Register a state change listener.
     */
    onStateChange(listener) {
        this.listeners.push(listener);
        return () => {
            const index = this.listeners.indexOf(listener);
            if (index >= 0) {
                this.listeners.splice(index, 1);
            }
        };
    }
    /**
     * Attempt to transition from allowed states to a new state.
     * @param fromStates Allowed states to transition from
     * @param toState Target state
     * @param action Optional action to execute during transition
     * @returns Promise that resolves to true if transition succeeded
     */
    async transition(fromStates, toState, action) {
        // Wait for any existing transition to complete
        if (this.transitionLock) {
            await this.transitionLock.catch(() => {
                // Ignore errors from previous transitions
            });
        }
        // Create a new transition lock
        let resolve;
        let reject;
        this.transitionLock = new Promise((res, rej) => {
            resolve = res;
            reject = rej;
        });
        try {
            const allowedStates = Array.isArray(fromStates) ? fromStates : [fromStates];
            // Check if transition is allowed
            if (!allowedStates.includes(this.currentState)) {
                resolve();
                return false;
            }
            // If already in target state, no need to transition
            if (this.currentState === toState) {
                resolve();
                return true;
            }
            const oldState = this.currentState;
            // Execute action if provided
            if (action) {
                await action();
            }
            // Update state
            this.currentState = toState;
            // Notify listeners
            for (const listener of this.listeners) {
                try {
                    listener(oldState, toState);
                }
                catch (error) {
                    // Don't let listener errors break the state machine
                    console.error('Error in state change listener:', error);
                }
            }
            resolve();
            return true;
        }
        catch (error) {
            reject(error);
            throw error;
        }
        finally {
            // Clear the transition lock
            this.transitionLock = null;
        }
    }
    /**
     * Force a state change without checking allowed transitions.
     * Use with caution - mainly for error recovery.
     */
    async forceState(newState) {
        if (this.transitionLock) {
            await this.transitionLock.catch(() => { });
        }
        const oldState = this.currentState;
        this.currentState = newState;
        // Notify listeners
        for (const listener of this.listeners) {
            try {
                listener(oldState, newState);
            }
            catch (error) {
                console.error('Error in state change listener:', error);
            }
        }
    }
}
exports.ConnectionStateMachine = ConnectionStateMachine;
