// @reclaimprotocol/attestor-core/lib/utils/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttestorError = void 0;
const api_1 = require("../proto/api");
/**
 * Represents an error that can be thrown by the Attestor Core
 * or server. Provides a code, and optional data
 * to pass along with the error.
 */
class AttestorError extends Error {
    constructor(code, message, data) {
        super(message);
        this.code = code;
        this.message = message;
        this.data = data;
        this.name = 'AttestorError';
    }
    /**
     * Encodes the error as a ErrorData
     * protobuf message
     */
    toProto() {
        return api_1.ErrorData.create({
            code: api_1.ErrorCode[this.code],
            message: this.message,
            data: JSON.stringify(this.data)
        });
    }
    static fromProto(data = api_1.ErrorData.fromJSON({})) {
        return new AttestorError(api_1.ErrorCode[data.code], data.message, data.data ? JSON.parse(data.data) : undefined);
    }
    static fromError(err) {
        if (err instanceof AttestorError) {
            return err;
        }
        return new AttestorError('ERROR_INTERNAL', err.message);
    }
    static badRequest(message, data) {
        return new AttestorError('ERROR_BAD_REQUEST', message, data);
    }
}
exports.AttestorError = AttestorError;
