// @reclaimprotocol/attestor-core/lib/utils/resource-monitor.js
"use strict";
/**
 * Resource monitor for tracking resources and preventing leaks.
 * Provides debugging capabilities and resource management.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceMonitor = void 0;
class ResourceMonitor {
    constructor(logger) {
        this.resources = new Map();
        this.logger = logger;
    }
    /**
     * Get or create the singleton ResourceMonitor instance.
     */
    static getInstance(logger) {
        if (!ResourceMonitor.instance) {
            ResourceMonitor.instance = new ResourceMonitor(logger);
        }
        return ResourceMonitor.instance;
    }
    /**
     * Track a new resource.
     */
    trackResource(id, type, metadata) {
        const resource = {
            id,
            type,
            createdAt: new Date(),
            metadata
        };
        this.resources.set(id, resource);
        this.logger.trace({ id, type, metadata }, 'Resource tracked');
    }
    /**
     * Untrack a resource.
     */
    untrackResource(id) {
        const resource = this.resources.get(id);
        if (resource) {
            this.resources.delete(id);
            this.logger.trace({ id, type: resource.type }, 'Resource untracked');
            return true;
        }
        return false;
    }
    /**
     * Check if a resource is tracked.
     */
    isTracked(id) {
        return this.resources.has(id);
    }
    /**
     * Get information about a specific resource.
     */
    getResource(id) {
        return this.resources.get(id);
    }
    /**
     * Get all tracked resources.
     */
    getAllResources() {
        return Array.from(this.resources.values());
    }
    /**
     * Get resources by type.
     */
    getResourcesByType(type) {
        return Array.from(this.resources.values()).filter(r => r.type === type);
    }
    /**
     * Get resources older than a specific age.
     */
    getResourcesOlderThan(ageMs) {
        const cutoff = new Date(Date.now() - ageMs);
        return Array.from(this.resources.values()).filter(r => r.createdAt < cutoff);
    }
    /**
     * Log resource statistics.
     */
    logStats() {
        const byType = new Map();
        for (const resource of this.resources.values()) {
            byType.set(resource.type, (byType.get(resource.type) || 0) + 1);
        }
        const stats = Object.fromEntries(byType);
        this.logger.info({
            totalResources: this.resources.size,
            byType: stats
        }, 'Resource statistics');
    }
    /**
     * Clear all tracked resources (for testing/cleanup).
     */
    clear() {
        this.resources.clear();
    }
    /**
     * Reset the singleton instance (for testing).
     */
    static reset() {
        ResourceMonitor.instance = null;
    }
}
exports.ResourceMonitor = ResourceMonitor;
ResourceMonitor.instance = null;
