// @reclaimprotocol/attestor-core/lib/window-rpc/window-rpc-zk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ALL_ENC_ALGORITHMS = void 0;
exports.makeWindowRpcZkOperator = makeWindowRpcZkOperator;
exports.makeWindowRpcOprfOperator = makeWindowRpcOprfOperator;
const utils_1 = require("../utils");
const utils_2 = require("../window-rpc/utils");
exports.ALL_ENC_ALGORITHMS = [
    'aes-256-ctr',
    'aes-128-ctr',
    'chacha20',
];
/**
 * The goal of this RPC operator is if the attestor client
 * is running in a WebView, it can call the native
 * application to perform the ZK operations
 */
function makeWindowRpcZkOperator(algorithm, bridge, zkEngine = 'snarkjs') {
    return {
        async generateWitness(input) {
            const operator = await (0, utils_1.makeDefaultZkOperator)(algorithm, zkEngine, utils_1.logger);
            return operator.generateWitness(input);
        },
        groth16Prove(input) {
            return callFn({ fn: 'groth16Prove', args: [input] });
        },
        groth16Verify(publicSignals, proof) {
            return callFn({ fn: 'groth16Verify', args: [publicSignals, proof] });
        },
    };
    function callFn(opts) {
        const id = (0, utils_2.generateRpcRequestId)();
        const waitForRes = (0, utils_2.waitForResponse)('executeZkFunctionV3', id, bridge);
        bridge.send({
            type: 'executeZkFunctionV3',
            id,
            request: opts,
            module: 'attestor-core'
        });
        return waitForRes;
    }
}
/**
 * The goal of this RPC operator is if the attestor client
 * is running in a WebView, it can call the native
 * application to perform the OPRF operations
 */
function makeWindowRpcOprfOperator(algorithm, bridge, zkEngine = 'snarkjs') {
    return {
        async generateWitness(input) {
            const operator = await (0, utils_1.makeDefaultZkOperator)(algorithm, zkEngine, utils_1.logger);
            return operator.generateWitness(input);
        },
        groth16Prove(input) {
            return callFn({ fn: 'groth16Prove', args: [input] });
        },
        groth16Verify(publicSignals, proof) {
            return callFn({ fn: 'groth16Verify', args: [publicSignals, proof] });
        },
        generateThresholdKeys(total, threshold) {
            return callFn({ fn: 'generateThresholdKeys', args: [total, threshold] });
        },
        generateOPRFRequestData(data, domainSeparator) {
            return callFn({ fn: 'generateOPRFRequestData', args: [data, domainSeparator] });
        },
        finaliseOPRF(serverPublicKey, request, responses) {
            return callFn({ fn: 'finaliseOPRF', args: [serverPublicKey, request, responses] });
        },
        evaluateOPRF(serverPrivateKey, request) {
            return callFn({ fn: 'evaluateOPRF', args: [serverPrivateKey, request] });
        },
    };
    function callFn(opts) {
        const id = (0, utils_2.generateRpcRequestId)();
        const waitForRes = (0, utils_2.waitForResponse)('executeOprfFunctionV3', id, bridge);
        bridge.send({
            type: 'executeOprfFunctionV3',
            id,
            request: opts,
            module: 'attestor-core'
        });
        return waitForRes;
    }
}
