// @reclaimprotocol/attestor-core/lib/tests/describe-with-server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.describeWithServer = void 0;
const client_socket_1 = require("../client/utils/client-socket");
const config_1 = require("../config");
const server_1 = require("../server");
const mock_provider_server_1 = require("../tests/mock-provider-server");
const mocks_1 = require("../tests/mocks");
const utils_1 = require("../tests/utils");
const utils_2 = require("../utils");
/**
 * Boots up a attestor server, a mock https server,
 * and a client that is renewed for each test.
 */
const describeWithServer = (name, fn) => describe(name, () => {
    let wsServer;
    let wsServerUrl;
    let privateKeyHex;
    let client;
    const wsServerPort = (0, utils_1.getRandomPort)();
    const httpsServerPort = (0, utils_1.getRandomPort)();
    const mockHttpsServer = (0, mock_provider_server_1.createMockServer)(httpsServerPort);
    beforeAll(async () => {
        wsServer = await (0, server_1.createServer)(wsServerPort);
        wsServerUrl = `ws://localhost:${wsServerPort}${config_1.WS_PATHNAME}`;
    });
    afterAll(() => {
        wsServer.close();
        mockHttpsServer.server.close();
    });
    beforeEach(async () => {
        mocks_1.SPY_PREPARER.mockClear();
        privateKeyHex = (0, utils_1.randomPrivateKey)();
        client = new client_socket_1.AttestorClient({
            logger: utils_2.logger.child({ client: 1 }),
            url: wsServerUrl,
        });
        await client.waitForInit();
    });
    afterEach(async () => {
        await client.terminateConnection();
    });
    fn({
        getClientOnServer,
        get client() {
            return client;
        },
        get privateKeyHex() {
            return privateKeyHex;
        },
        get serverUrl() {
            return wsServerUrl;
        },
        mockHttpsServer,
        mockhttpsServerPort: httpsServerPort,
    });
    function getClientOnServer() {
        var _a;
        const serverSockets = [...wsServer.clients.values()];
        return (_a = serverSockets.at(-1)) === null || _a === void 0 ? void 0 : _a.serverSocket;
    }
});
exports.describeWithServer = describeWithServer;
