// @reclaimprotocol/attestor-core/lib/tests/test.rpc-tunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("../client");
const make_rpc_tcp_tunnel_1 = require("../client/tunnels/make-rpc-tcp-tunnel");
const make_rpc_tls_tunnel_1 = require("../client/tunnels/make-rpc-tls-tunnel");
const describe_with_server_1 = require("../tests/describe-with-server");
const utils_1 = require("../tests/utils");
const utils_2 = require("../utils");
(0, describe_with_server_1.describeWithServer)('RPC Tunnel', opts => {
    const { mockHttpsServer, getClientOnServer } = opts;
    let client;
    beforeEach(async () => {
        client = opts.client;
    });
    afterEach(async () => {
        await client.terminateConnection();
    });
    it('should connect to a server via RPC tunnel', async () => {
        // setup tunnel for listening & then
        // connect to it via RPC
        const tunnel = await (0, make_rpc_tcp_tunnel_1.makeRpcTcpTunnel)({ tunnelId: 1, client });
        await client.rpc('createTunnel', {
            id: 1,
            host: 'localhost',
            port: opts.mockhttpsServerPort,
            geoLocation: ''
        });
        const ws = getClientOnServer();
        const socketTunnel = ws === null || ws === void 0 ? void 0 : ws.tunnels[1];
        expect(socketTunnel).toBeTruthy();
        await tunnel.close();
        // check that the server actually closed the tunnel
        // upon our request
        await expect(socketTunnel === null || socketTunnel === void 0 ? void 0 : socketTunnel.write(Buffer.from('hello'))).rejects.toMatchObject({
            code: 'ERR_STREAM_DESTROYED'
        });
    });
    describe('TLS', () => {
        it('should do a TLS handshake via RPC tunnel', async () => {
            const ws = getClientOnServer();
            const tunnel = await (0, make_rpc_tls_tunnel_1.makeRpcTlsTunnel)({
                request: {
                    id: 1,
                    host: 'localhost',
                    port: opts.mockhttpsServerPort,
                },
                tlsOpts: {
                    verifyServerCertificate: false,
                },
                logger: client.logger,
                connect(initMessages) {
                    client.sendMessage(...initMessages)
                        .catch(() => { });
                    // ensure that the client hello message
                    // was sent to the server along the
                    // "createTunnel" request -- that saves
                    // us a round-trip
                    expect(initMessages[1].tunnelMessage)
                        .toBeTruthy();
                    return client;
                },
            });
            expect(ws === null || ws === void 0 ? void 0 : ws.tunnels[1]).toBeTruthy();
            await tunnel.close();
        });
        it('should setup a 0-RTT TLS connection', async () => {
            let client2;
            const tunnel = await (0, make_rpc_tls_tunnel_1.makeRpcTlsTunnel)({
                request: {
                    id: 1,
                    host: 'localhost',
                    port: opts.mockhttpsServerPort,
                },
                tlsOpts: {
                    verifyServerCertificate: false,
                },
                logger: client.logger,
                connect(initMessages) {
                    client2 = new client_1.AttestorClient({
                        url: opts.serverUrl,
                        logger: utils_2.logger.child({ client: 2 }),
                        initMessages
                    });
                    return client2;
                },
            });
            await tunnel.close();
            await (client2 === null || client2 === void 0 ? void 0 : client2.terminateConnection());
        });
        it('should gracefully handle a TLS disconnection alert', async () => {
            let socket;
            mockHttpsServer.server.once('secureConnection', s => {
                socket = s;
            });
            let closeResolve;
            await (0, make_rpc_tls_tunnel_1.makeRpcTlsTunnel)({
                request: {
                    id: 1,
                    host: 'localhost',
                    port: opts.mockhttpsServerPort,
                },
                tlsOpts: {
                    verifyServerCertificate: false,
                },
                logger: client.logger,
                connect(initMessages) {
                    client.sendMessage(...initMessages)
                        .catch(() => { });
                    return client;
                },
                onClose(err) {
                    closeResolve === null || closeResolve === void 0 ? void 0 : closeResolve(err);
                },
            });
            await (0, utils_1.delay)(100);
            expect(socket).toBeTruthy();
            socket === null || socket === void 0 ? void 0 : socket.end();
            const err = await new Promise((resolve) => {
                closeResolve = resolve;
            });
            // since it was a graceful close, there should be no error
            expect(err).toBeUndefined();
        });
        it('should handle TLS handshake errors', async () => {
            await expect((0, make_rpc_tls_tunnel_1.makeRpcTlsTunnel)({
                request: {
                    id: 1,
                    host: 'localhost',
                    port: opts.mockhttpsServerPort,
                },
                tlsOpts: {
                    applicationLayerProtocols: [
                        'invalid-protocol'
                    ]
                },
                logger: client.logger,
                connect(initMessages) {
                    client.sendMessage(...initMessages)
                        .catch(() => { });
                    return client;
                },
            })).rejects.toMatchObject({
                message: /NO_APPLICATION_PROTOCOL/
            });
        });
        it('should handle tunnel creation errors', async () => {
            await expect((0, make_rpc_tls_tunnel_1.makeRpcTlsTunnel)({
                request: {
                    id: 1,
                    host: 'localhost',
                    port: opts.mockhttpsServerPort,
                    // invalid geo location
                    geoLocation: 'XZ'
                },
                tlsOpts: {
                    applicationLayerProtocols: [
                        'invalid-protocol'
                    ]
                },
                logger: client.logger,
                connect(initMessages) {
                    client.sendMessage(...initMessages)
                        .catch(() => { });
                    return client;
                },
            })).rejects.toMatchObject({
                message: /NO_APPLICATION_PROTOCOL/
            });
        });
    });
});
