// @reclaimprotocol/attestor-core/lib/tests/test.auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ethers_1 = require("ethers");
const client_1 = require("../client");
const config_1 = require("../config");
const providers_1 = require("../providers");
const server_1 = require("../server");
const mock_provider_server_1 = require("../tests/mock-provider-server");
const utils_1 = require("../tests/utils");
const utils_2 = require("../utils");
describe('Authentication Tests', () => {
    const authKp = ethers_1.ethers.Wallet.createRandom();
    let wsServer;
    let wsServerUrl;
    let claimUrl;
    let privateKeyHex;
    const wsServerPort = (0, utils_1.getRandomPort)();
    const httpsServerPort = (0, utils_1.getRandomPort)();
    const mockHttpsServer = (0, mock_provider_server_1.createMockServer)(httpsServerPort);
    beforeAll(async () => {
        wsServer = await (0, server_1.createServer)(wsServerPort);
        wsServerUrl = `ws://localhost:${wsServerPort}${config_1.WS_PATHNAME}`;
        process.env.AUTHENTICATION_PUBLIC_KEY = authKp.publicKey;
        claimUrl = `https://localhost:${httpsServerPort}/me`;
        privateKeyHex = (0, utils_1.randomPrivateKey)();
        // we need to disable certificate verification
        // for testing purposes
        providers_1.providers.http.additionalClientOptions = {
            verifyServerCertificate: false
        };
    });
    afterAll(() => {
        delete process.env.AUTHENTICATION_PUBLIC_KEY;
        wsServer.close();
        mockHttpsServer.server.close();
    });
    it('should fail to create a claim w/o authentication', async () => {
        await expect(createClaim(undefined)).rejects.toMatchObject({
            message: 'User must be authenticated'
        });
    });
    it('should block claim creation if host not in whitelist', async () => {
        const auth = await (0, utils_2.createAuthRequest)({ id: '1234', hostWhitelist: ['api.abcd.com'] }, authKp.privateKey);
        await expect(createClaim(auth)).rejects.toMatchObject({
            message: 'Host \"localhost\" not allowed by auth request'
        });
    });
    it('should create claim after authentication', async () => {
        const auth = await (0, utils_2.createAuthRequest)({ id: '1234', hostWhitelist: ['localhost'] }, authKp.privateKey);
        await createClaim(auth);
    });
    function createClaim(authRequest) {
        const user = 'testing-123';
        return (0, client_1.createClaimOnAttestor)({
            name: 'http',
            params: {
                url: claimUrl,
                method: 'GET',
                responseRedactions: [],
                responseMatches: [
                    {
                        type: 'contains',
                        value: `${user}@mock.com`
                    }
                ]
            },
            secretParams: {
                authorisationHeader: `Bearer ${user}`
            },
            ownerPrivateKey: privateKeyHex,
            client: { url: wsServerUrl, authRequest }
        });
    }
});
