// @reclaimprotocol/attestor-core/lib/tests/test.tcp-tunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tls_1 = require("@reclaimprotocol/tls");
const server_1 = require("../server");
const utils_1 = require("../utils");
const DEMO_GEO_LOCATIONS = ['in', 'us'];
jest.setTimeout(15000);
describe.skip('TCP Tunnel', () => {
    it.each([...DEMO_GEO_LOCATIONS, 'none'])('should generate a session using a geoLocation (%s)', async (geoLocation) => {
        const resParser = (0, utils_1.makeHttpResponseParser)();
        let resolvePromise;
        let rejectPromise;
        const session = await (0, server_1.makeTcpTunnel)({
            host: 'lumtest.com',
            port: 80,
            geoLocation: geoLocation === 'none' ? '' : geoLocation,
            logger: utils_1.logger,
            onClose(err) {
                rejectPromise === null || rejectPromise === void 0 ? void 0 : rejectPromise(err || new Error('session closed'));
            },
            onMessage(data) {
                resParser.onChunk(data);
                if (resParser.res.complete) {
                    resolvePromise === null || resolvePromise === void 0 ? void 0 : resolvePromise();
                }
            },
        });
        const waitForRes = new Promise((resolve, reject) => {
            resolvePromise = resolve;
            rejectPromise = reject;
        });
        const str = 'GET /myip.json HTTP/1.1\r\nHost: lumtest.com\r\n\r\n';
        await session.write((0, tls_1.strToUint8Array)(str));
        await waitForRes;
        await session.close();
        expect(resParser.res.statusCode).toBe(200);
        const resBody = (0, tls_1.uint8ArrayToStr)(resParser.res.body);
        const resJson = JSON.parse(resBody);
        if (geoLocation === 'none') {
            return;
        }
        expect(resJson.country).toBe(geoLocation.toUpperCase());
    });
    it('should gracefully fail an invalid geoLocation', async () => {
        await expect((0, server_1.makeTcpTunnel)({
            host: 'lumtest.com',
            port: 80,
            geoLocation: 'xz',
            logger: utils_1.logger,
        })).rejects.toMatchObject({
            message: /failed with status code: 400/
        });
    });
    it('should connect to restricted server', async () => {
        const session = await (0, server_1.makeTcpTunnel)({
            host: 'servicos.acesso.gov.br',
            port: 80,
            geoLocation: 'US',
            logger: utils_1.logger,
        });
        await session.close();
    });
});
