// @reclaimprotocol/attestor-core/lib/server/handlers/createClaimOnChain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClaimOnChain = void 0;
const contracts_1 = require("../../avs/utils/contracts");
const tasks_1 = require("../../avs/utils/tasks");
const utils_1 = require("../../utils");
const env_1 = require("../../utils/env");
const ACCEPT_CLAIM_PAYMENT_REQUESTS = (0, env_1.getEnvVariable)('ACCEPT_CLAIM_PAYMENT_REQUESTS') === '1';
const createClaimOnChain = async ({ chainId: chainIdNum, jsonCreateClaimRequest, requestSignature }) => {
    if (!ACCEPT_CLAIM_PAYMENT_REQUESTS) {
        throw new utils_1.AttestorError('ERROR_PAYMENT_REFUSED', 'Payment requests are not accepted at this time');
    }
    const chainId = chainIdNum.toString();
    const { wallet } = (0, contracts_1.getContracts)(chainId.toString());
    const request = JSON.parse(jsonCreateClaimRequest);
    const { task, tx } = await (0, tasks_1.createNewClaimRequestOnChain)({
        request,
        owner: request.owner,
        payer: wallet,
        chainId,
        requestSignature: requestSignature
    });
    const plainTask = (0, utils_1.ethersStructToPlainObject)(task);
    return {
        txHash: tx.transactionHash,
        taskIndex: task.taskIndex,
        jsonTask: JSON.stringify(plainTask)
    };
};
exports.createClaimOnChain = createClaimOnChain;
