// @reclaimprotocol/attestor-core/lib/avs/utils/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContracts = getContracts;
exports.initialiseContracts = initialiseContracts;
const ethers_1 = require("ethers");
const avsDirectoryABI_1 = require("../../avs/abis/avsDirectoryABI");
const delegationABI_1 = require("../../avs/abis/delegationABI");
const registryABI_1 = require("../../avs/abis/registryABI");
const config_1 = require("../../avs/config");
const contracts_1 = require("../../avs/contracts");
const configs = {};
/**
 * get the contracts for the given chain ID
 */
function getContracts(chainId = config_1.SELECTED_CHAIN_ID) {
    const config = config_1.CHAIN_CONFIGS[chainId];
    if (!config) {
        throw new Error(`No config found for chain ID: ${chainId}`);
    }
    configs[chainId] || (configs[chainId] = initialiseContracts(config));
    return configs[chainId];
}
function initialiseContracts({ rpcUrl, stakeRegistryAddress, avsDirectoryAddress, contractAddress, delegationManagerAddress, }, privateKey = config_1.PRIVATE_KEY) {
    const provider = new ethers_1.ethers.providers.JsonRpcProvider(rpcUrl);
    const wallet = privateKey
        ? new ethers_1.ethers.Wallet(privateKey, provider)
        : undefined;
    return {
        provider,
        wallet,
        delegationManager: new ethers_1.ethers.Contract(delegationManagerAddress, delegationABI_1.delegationABI, wallet || provider),
        // eslint-disable-next-line camelcase
        contract: contracts_1.ReclaimServiceManager__factory.connect(contractAddress, wallet || provider),
        registryContract: new ethers_1.ethers.Contract(stakeRegistryAddress, registryABI_1.registryABI, wallet || provider),
        avsDirectory: new ethers_1.ethers.Contract(avsDirectoryAddress, avsDirectoryABI_1.avsDirectoryABI, wallet || provider),
    };
}
