// @reclaimprotocol/attestor-core/lib/avs/utils/register.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerOperator = registerOperator;
const ethers_1 = require("ethers");
const config_1 = require("../../avs/config");
const contracts_1 = require("../../avs/utils/contracts");
const utils_1 = require("../../utils");
/**
 * Registers the operator on the chain, if required.
 * If already registered -- will just pass through
 */
async function registerOperator({ logger = utils_1.logger, chainId = config_1.SELECTED_CHAIN_ID, wallet = (0, contracts_1.getContracts)(chainId).wallet, reclaimRpcUrl = config_1.RECLAIM_PUBLIC_URL } = {}) {
    const contracts = (0, contracts_1.getContracts)(chainId);
    const delegationManager = contracts.delegationManager
        .connect(wallet);
    const avsDirectory = contracts.avsDirectory
        .connect(wallet);
    const contract = contracts.contract
        .connect(wallet);
    const registryContract = contracts.registryContract
        .connect(wallet);
    const addr = await wallet.address;
    try {
        const tx1 = await delegationManager
            .registerAsOperator({
            earningsReceiver: addr,
            delegationApprover: '0x0000000000000000000000000000000000000000',
            stakerOptOutWindowBlocks: 0
        }, '');
        await tx1.wait();
        logger.info('operator registered on DM successfully');
    }
    catch (err) {
        if (!err.message.includes('operator has already registered')) {
            throw err;
        }
        logger.info('Operator already registered on EL');
    }
    const salt = ethers_1.ethers.utils.hexlify(ethers_1.ethers.utils.randomBytes(32));
    // Example expiry, 1 hour from now
    const expiry = Math.floor(Date.now() / 1000) + 3600;
    // Define the output structure
    const operatorSignature = {
        expiry: expiry,
        salt: salt,
        signature: ''
    };
    // Calculate the digest hash using the avsDirectory's method
    const digestHash = await avsDirectory
        .calculateOperatorAVSRegistrationDigestHash(addr, contract.address, salt, expiry);
    // Sign the digest hash with the operator's private key
    const signingKey = new ethers_1.ethers.utils.SigningKey(wallet.privateKey);
    const signature = signingKey.signDigest(digestHash);
    // Encode the signature in the required format
    operatorSignature.signature = ethers_1.ethers.utils.joinSignature(signature);
    logger.info('operator signature generated successfully');
    if (!(await registryContract.operatorRegistered(addr))) {
        const tx2 = await registryContract
            .registerOperatorWithSignature(addr, operatorSignature);
        await tx2.wait();
        logger.info('operator registered on AVS successfully');
    }
    else {
        logger.info('Operator already registered on AVS');
    }
    const existingMetadata = await contract.getMetadataForOperator(addr)
        .catch(err => {
        if (err.message.includes('Operator not found')) {
            return undefined;
        }
        throw err;
    });
    const metadata = { addr, url: reclaimRpcUrl };
    if ((existingMetadata === null || existingMetadata === void 0 ? void 0 : existingMetadata.addr) === metadata.addr
        && (existingMetadata === null || existingMetadata === void 0 ? void 0 : existingMetadata.url) === metadata.url) {
        logger.info('operator metadata already up to date');
        return;
    }
    await contract.updateOperatorMetadata(metadata);
    logger.info({ metadata }, 'operator metadata updated successfully');
}
