// @reclaimprotocol/attestor-core/lib/avs/client/create-claim-on-avs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClaimOnAvs = createClaimOnAvs;
const ethers_1 = require("ethers");
const config_1 = require("../../avs/config");
const contracts_1 = require("../../avs/utils/contracts");
const tasks_1 = require("../../avs/utils/tasks");
const client_1 = require("../../client");
const utils_1 = require("../../utils");
const EMPTY_CLAIM_USER_ID = ethers_1.ethers.utils.hexlify(new Uint8Array(32));
/**
 * Creates a Reclaim claim on the AVS chain.
 */
async function createClaimOnAvs({ onStep, createClaimOnAttestor = client_1.createClaimOnAttestor, chainId = config_1.SELECTED_CHAIN_ID, payer, ...opts }) {
    var _a, _b, _c, _d, _e;
    const { logger = utils_1.logger, ownerPrivateKey, name, params, context, } = opts;
    const { contract, wallet } = (0, contracts_1.initialiseContracts)(config_1.CHAIN_CONFIGS[chainId], ownerPrivateKey);
    logger.info({ owner: wallet.address, contract: contract.address }, 'creating claim');
    const arg = await requestClaimCreation();
    logger.info({
        index: arg.taskIndex,
        operators: arg.task.operators.length,
    }, 'task created, collecting claim signatures...');
    onStep === null || onStep === void 0 ? void 0 : onStep({ type: 'taskCreated', data: arg });
    const responses = [];
    const timestampS = +arg.task.createdAt.toString();
    for (const op of arg.task.operators) {
        const res = await createClaimOnAttestor({
            ...opts,
            client: { url: op.url },
            timestampS,
            onStep: (step) => (onStep === null || onStep === void 0 ? void 0 : onStep({
                type: 'attestorStep',
                data: {
                    operatorAddress: op.addr,
                    step,
                },
            })),
            logger: logger.child({ operator: op.addr }),
        });
        const signature = (_a = res.signatures) === null || _a === void 0 ? void 0 : _a.claimSignature;
        if (!signature) {
            throw new Error('INTERNAL: Claim signature not generated');
        }
        const diff = getClaimRequestDifference((_b = res.request) === null || _b === void 0 ? void 0 : _b.data, res.claim);
        if (diff) {
            throw new utils_1.AttestorError('ERROR_INVALID_CLAIM', `Claim request does not match the claim res data: ${diff}`, {
                diff,
                request: (_d = (_c = res.request) === null || _c === void 0 ? void 0 : _c.data) === null || _d === void 0 ? void 0 : _d[diff],
                claim: (_e = res.claim) === null || _e === void 0 ? void 0 : _e[diff]
            });
        }
        responses.push(res);
        logger.info({ operator: op.addr }, 'signature generated');
        onStep === null || onStep === void 0 ? void 0 : onStep({
            type: 'attestorDone',
            data: {
                task: arg,
                responsesDone: responses,
            },
        });
    }
    const rslt = await completeTask();
    logger.info({ tx: rslt.txHash, task: arg.taskIndex }, 'claim submitted & validated');
    return { ...rslt, claimData: responses[0].claim };
    async function requestClaimCreation() {
        const request = {
            provider: name,
            // blank for now -- till we figure out the right
            // algorithm for this
            claimUserId: EMPTY_CLAIM_USER_ID,
            claimHash: (0, utils_1.getIdentifierFromClaimInfo)({
                provider: name,
                parameters: (0, utils_1.canonicalStringify)(params),
                context: context
                    ? (0, utils_1.canonicalStringify)(context)
                    : '',
            }),
            owner: wallet.address,
            requestedAt: (0, utils_1.unixTimestampSeconds)()
        };
        if (!payer) {
            const wallet = new ethers_1.Wallet(ownerPrivateKey, contract.provider);
            const { task } = await (0, tasks_1.createNewClaimRequestOnChain)({
                request,
                payer: wallet,
                owner: wallet,
                chainId
            });
            return task;
        }
        const requestSignature = await (0, tasks_1.signClaimRequest)(request, wallet, chainId);
        const client = (0, client_1.getAttestorClientFromPool)(payer.attestor);
        await client.waitForInit();
        const rslt = await client.rpc('createClaimOnChain', {
            chainId: +chainId,
            jsonCreateClaimRequest: JSON.stringify(request),
            requestSignature
        });
        return JSON.parse(rslt.jsonTask);
    }
    async function completeTask() {
        var _a;
        const data = {
            task: arg.task,
            signatures: responses
                .map(res => {
                var _a;
                return (ethers_1.ethers.utils.hexlify((_a = res.signatures) === null || _a === void 0 ? void 0 : _a.claimSignature));
            }),
        };
        if (!payer) {
            const tx = await contract.taskCompleted(data, arg.taskIndex);
            const rslt = await tx.wait();
            // check task created event was emitted
            const ev = (_a = rslt.events) === null || _a === void 0 ? void 0 : _a[0];
            return {
                object: ev === null || ev === void 0 ? void 0 : ev.args,
                txHash: rslt.transactionHash
            };
        }
        const client = (0, client_1.getAttestorClientFromPool)(payer.attestor);
        await client.waitForInit();
        const rslt = await client.rpc('completeClaimOnChain', {
            chainId: +chainId,
            taskIndex: arg.taskIndex,
            completedTaskJson: JSON.stringify(data)
        });
        const object = JSON.parse(rslt.taskCompletedObjectJson);
        return { object, txHash: rslt.txHash };
    }
}
function getClaimRequestDifference(a, b) {
    if (a.provider !== b.provider) {
        return 'provider';
    }
    if (a.context !== b.context) {
        return 'context';
    }
    if (a.parameters !== b.parameters) {
        return 'parameters';
    }
    if (a.timestampS !== b.timestampS) {
        return 'timestampS';
    }
}
